README
NAME

    Net::Async::Redis - talk to Redis servers via IO::Async

SYNOPSIS

        use Net::Async::Redis;
        use Future::AsyncAwait;
        use IO::Async::Loop;
        my $loop = IO::Async::Loop->new;
        $loop->add(my $redis = Net::Async::Redis->new);
        (async sub {
         await $redis->connect;
         my $value = await $redis->get('some_key');
         $value ||= await $redis->set(some_key => 'some_value');
         print "Value: $value";
        })->()->get;
    
        # You can also use ->then chaining, see L<Future> for more details
        $redis->connect->then(sub {
            $redis->get('some_key')
        })->then(sub {
            my $value = shift;
            return Future->done($value) if $value;
            $redis->set(some_key => 'some_value')
        })->on_done(sub {
            print "Value: " . shift;
        })->get;
    
        # ... or with Future::AsyncAwait
        await $redis->connect;
        my $value = await $redis->get('some_key');
        $value ||= await $redis->set(some_key => 'some_value');
        print "Value: $value";

DESCRIPTION

    Provides client access for dealing with Redis servers.

    See Net::Async::Redis::Commands for the full list of commands, this
    list is autogenerated from the official documentation here:

    https://redis.io/commands

    This is intended to be a near-complete low-level client module for
    asynchronous Redis support. See Net::Async::Redis::Server for a
    (limited) Perl server implementation. It is an unofficial Perl port and
    not endorsed by the Redis server maintainers in any way.

 Supported features

    Current features include:

      * all commands <https://redis.io/commands> as of 6.0 (May 2020), see
      https://redis.io/commands for the methods and parameters

      * pub/sub support <https://redis.io/topics/pubsub>, see
      "Subscriptions"

      * pipelining <https://redis.io/topics/pipelining>, see "Pipelining"

      * transactions <https://redis.io/topics/transactions>, see
      "Transactions"

      * streams <https://redis.io/topics/streams-intro> and consumer
      groups, see "Streams"

 Connecting

    As with any other IO::Async::Notifier-based module, you'll need to add
    this to an IO::Async::Loop:

        my $loop = IO::Async::Loop->new;
        $loop->add(
            my $redis = Net::Async::Redis->new
        );

    then connect to the server:

        $redis->connect
            ->then(sub {
                # You could achieve a similar result by passing client_name in
                # constructor or ->connect parameters
                $redis->client_setname("example client")
            })->get;

 Key-value handling

    One of the most common Redis scenarios is as a key/value store. The
    "get" and "set" methods are typically used here:

     $redis->set(some_key => 'some value')
      ->then(sub {
       $redis->get('some_key')
      })->on_done(sub {
       my ($value) = @_;
       print "Read back value [$value]\n";
      })->retain;

    See the next section for more information on what these methods are
    actually returning.

 Requests and responses

    Requests are implemented as methods on the Net::Async::Redis object.
    These typically return a Future which will resolve once ready:

        my $future = $redis->incr("xyz")
            ->on_done(sub {
                print "result of increment was " . shift . "\n"
            });

    For synchronous code, call ->get on that Future:

        print "Database has " . $redis->dbsize->get . " total keys\n";

    This means you can end up with ->get being called on the result of
    ->get, note that these are two very different methods:

     $redis
      ->get('some key') # this is being called on $redis, and is issuing a GET request
      ->get # this is called on the returned Future, and blocks until the value is ready

    Typical async code would not be expected to use the "get" in Future
    method extensively; often only calling it in one place at the top level
    in the code.

 Error handling

    Since Future is used for deferred results, failure is indicated by a
    failing Future with failure category of redis.

    The "catch" in Future feature may be useful for handling these:

     $redis->lpush(key => $value)
         ->catch(
             redis => sub { warn "probably an incorrect type, cannot push value"; Future->done }
         )->get;

    Note that this module uses Future::AsyncAwait internally.

METHODS

    NOTE: For a full list of the Redis methods supported by this module,
    please see Net::Async::Redis::Commands.

METHODS - Subscriptions

    See https://redis.io/topics/pubsub for more details on this topic.
    There's also more details on the internal implementation in Redis here:
    https://making.pusher.com/redis-pubsub-under-the-hood/.

 psubscribe

    Subscribes to a pattern.

    Example:

     # Subscribe to 'info::*' channels, i.e. any message
     # that starts with the C<info::> prefix, and prints them
     # with a timestamp.
     $redis_connection->psubscribe('info::*')
        ->then(sub {
            my $sub = shift;
            $sub->map('payload')
                ->each(sub {
                 print localtime . ' ' . $_ . "\n";
                })->retain
        })->get;
     # this will block until the subscribe is confirmed. Note that you can't publish on
     # a connection that's handling subscriptions due to Redis protocol restrictions.
     $other_redis_connection->publish('info::example', 'a message here')->get;

    Returns a Future which resolves to a Net::Async::Redis::Subscription
    instance.

 subscribe

    Subscribes to one or more channels.

    Returns a Future which resolves to a Net::Async::Redis::Subscription
    instance.

    Example:

     # Subscribe to 'notifications' channel,
     # print the first 5 messages, then unsubscribe
     $redis->subscribe('notifications')
        ->then(sub {
            my $sub = shift;
            $sub->events
                ->map('payload')
                ->take(5)
                ->say
                ->completed
        })->then(sub {
            $redis->unsubscribe('notifications')
        })->get

METHODS - Transactions

 multi

    Executes the given code in a Redis MULTI transaction.

    This will cause each of the requests to be queued on the server, then
    applied in a single atomic transaction.

    Note that the commands will resolve only after the transaction is
    committed: for example, when the "set" command is issued, Redis will
    return QUEUED. This information is not used as the result - we only
    pass through the immediate response if there was an error. The Future
    representing the response will be marked as done once the EXEC command
    is applied and we have the results back.

    Example:

     $redis->multi(sub {
      my $tx = shift;
      $tx->incr('some::key')->on_done(sub { print "Final value for incremented key was " . shift . "\n"; });
      $tx->set('other::key => 'test data')
     })->then(sub {
      my ($success, $failure) = @_;
      return Future->fail("Had $failure failures, expecting everything to succeed") if $failure;
      print "$success succeeded\m";
      return Future->done;
     })->retain;

METHODS - Clientside caching

    Enable clientside caching by passing a true value for
    client_side_caching_enabled in "configure" or "new". This is currently
    experimental, and only operates on "get" in Net::Async::Redis::Commands
    requests.

    See https://redis.io/topics/client-side-caching for more details on
    this feature.

METHODS - Generic

 keys

 watch_keyspace

    A convenience wrapper around the keyspace notifications API.

    Provides the necessary setup to establish a PSUBSCRIBE subscription on
    the __keyspace@*__ namespace, setting the configuration required for
    this to start emitting events, and then calls $code with each event.

    Note that this will switch the connection into pubsub mode, so it will
    no longer be available for any other activity.

    Resolves to a Ryu::Source instance.

 endpoint

    The string describing the remote endpoint.

 local_endpoint

    A string describing the local endpoint, usually host:port.

 connect

    Connects to the Redis server.

    Will use the "configure"d parameters if available, but as a convenience
    can be passed additional parameters which will then be applied as if
    you had called "configure" with those beforehand. This also means that
    they will be preserved for subsequent "connect" calls.

 connected

    Establishes a connection if needed, otherwise returns an
    immediately-available Future instance.

 on_message

    Called for each incoming message.

    Passes off the work to "handle_pubsub_message" or the next queue item,
    depending on whether we're dealing with subscriptions at the moment.

 stream

    Represents the IO::Async::Stream instance for the active Redis
    connection.

 pipeline_depth

    Number of requests awaiting responses before we start queuing. This
    defaults to an arbitrary value of 100 requests.

    Note that this does not apply when in transaction (MULTI) mode.

    See https://redis.io/topics/pipelining for more details on this
    concept.

METHODS - Deprecated

    This are still supported, but no longer recommended.

METHODS - Internal

 notify_close

    Called when the socket is closed.

 command_label

    Generate a label for the given command list.

 stream_read_len

    Defines the buffer size when reading from a Redis connection.

    Defaults to 1MB, reduce this if you're dealing with a lot of
    connections and want to minimise memory usage. Alternatively, if you're
    reading large amounts of data and spend too much time in needless
    epoll_wait calls, try a larger value.

 stream_write_len

    The buffer size when writing to Redis connections, in bytes. Defaults
    to 1MB.

    See "stream_read_len".

SEE ALSO

    Some other Redis implementations on CPAN:

      * Mojo::Redis2 - nonblocking, using the Mojolicious framework,
      actively maintained

      * MojoX::Redis - changelog mentions that this was obsoleted by
      Mojo::Redis, although there have been new versions released since
      then

      * RedisDB - another synchronous (blocking) implementation, handles
      pub/sub and autoreconnect

      * Cache::Redis - wrapper around RedisDB

      * Redis::Fast - wraps hiredis, faster than Redis

      * Redis::Jet - also XS-based, docs mention very early development
      stage but appears to support pipelining and can handle newer commands
      via ->command.

      * Redis - synchronous (blocking) implementation, handles pub/sub and
      autoreconnect

AUTHOR

    Tom Molesworth <TEAM@cpan.org>

CONTRIBUTORS

    With thanks to the following for contributing patches, bug reports,
    tests and feedback:

      * BINARY@cpan.org

      * PEVANS@cpan.org

      * @eyadof

      * Nael Alolwani

LICENSE

    Copyright Tom Molesworth and others 2015-2020. Licensed under the same
    terms as Perl itself.



Changes
Revision history for Net-Async-Redis

3.019     2022-02-16 15:42:23+08:00 Asia/Kuala_Lumpur
    [Bugs fixed]

    - the ->xgroup method was not available due to 7.0 changes, added back in

3.018     2022-02-01 11:58:52+08:00 Asia/Kuala_Lumpur
    [New features]

    - Added latest commands from Redis 7.0 release candidate

3.017     2021-10-18 00:15:14+08:00 Asia/Kuala_Lumpur
    [New features]

    - Added ->client_no_evict from latest Redis

3.016     2021-09-26 02:23:34+08:00 Asia/Kuala_Lumpur
    [New features]
    - Added readonly methods from latest Redis, e.g. `->sort_ro` and `->bitfield_ro`

3.015     2021-07-27 03:47:43+08:00 Asia/Kuala_Lumpur
    [Bugs fixed]
    - The CLUSTER INFO command generated warnings (RT138053, thanks to Nael for
    reporting)

3.014     2021-07-20 18:37:09+08:00 Asia/Kuala_Lumpur
    [New features]

    - ->expiretime and ->eval_ro from latest commands list

3.013     2021-04-27 10:28:39+08:00 Asia/Kuala_Lumpur
    No new features.

    Bugs fixed:

    - removed `dynamically` usage for `$log->{context}` since it can trigger a
    memory leak and also affects performance, see:
    https://github.com/binary-com/perl-Myriad/issues/117 for an example

3.012     2021-04-22 21:26:23+08:00 Asia/Kuala_Lumpur
    New features:

    - automatic connection and queuing is now more reliable, you should only
    need `await $redis->connected` in a few cases now
    - acess to client-side invalidation events, for implementing observables
    - enable keyspace notifications in cluster mode

    Bugs fixed:

    - various issues with client-side caching resolved
    - in cluster mode, XREAD could fail due to incorrect node lookup for key
    (Github #23, thanks to Marc Frank for the patch)
    - the stream read/write lengths were using the same internal key,
    so any changes to one value would affect the other as well
    - XINFO was mapping to the wrong key in cluster mode, thanks to Eyad for
    catching and fixing this

3.011     2021-02-03 03:43:05+08:00 Asia/Kuala_Lumpur
    New features:

    - ->failover added from latest commands list

3.010     2021-02-01 04:56:52+08:00 Asia/Kuala_Lumpur
    New features:

    - ->bootstrap will now fail if the cluster's nodes addresses are not initiated.
    - cluster->execute_command will now redirect the command to the correct node 
      if 'MOVED' error received.
    - latest commands from 6.2
    - this includes ->getdel, ->getex, ->zrandfield

3.009     2021-01-27 06:33:18+08:00 Asia/Kuala_Lumpur
    No new features.

    Bugs fixed:

    - `XGROUP` KEY_FINDER index was wrong.
    - `PUBLISH`/`SUBSCRIBE` were not in the KEY_FINDER index.
    - Net::Async::Redis::Cluster was ignoring the protocol compatibility layer.

3.008     2021-01-17 01:29:19+08:00 Asia/Kuala_Lumpur
    New features:

    - latest commands from 6.2
    - this includes ->xautoclaim, ->client_unpause and the `count => 'any'` option
    on various commands

3.007     2020-12-31 07:23:30+08:00 Asia/Kuala_Lumpur
    New features:

    - latest commands from 6.2
    - this includes ->geosearch, ->client_info, ->client_trackinginfo etc.

3.006     2020-12-04 05:33:59+08:00 Asia/Kuala_Lumpur
    New features:

    - latest commands from 6.0.9
    - ->copy is now supported (from latest Redis release)
    - stream handling now documents `MKSTREAM` and `NOMKSTREAM`,
    and updates the trimming to mention `~` (approximate) and `=`
    (exact) modes for removing old entries

3.005     2020-11-16 05:20:49+08:00 Asia/Kuala_Lumpur
    New features:

    - ->zdiff and ->zdiffstore commands from latest Redis release
    - provide RESP2/RESP3 compatibility when hashrefs option is not enabled,
    otherwise some commands such as ->xreadgroup can return inconsistent
    results between versions (thanks Eyad, Paul and Nael for the patch)
    - cluster mode now provides basic slot calculation caching, since this
    yields a small performance improvement

3.004     2020-11-04 12:40:40+08:00 Asia/Kuala_Lumpur
    No new features.

    Bugs fixed:

    - top-level hashes also now follow the `hashrefs` configuration option,
    this affects methods such as `->xreadgroup`.

3.003     2020-11-03 13:58:00+08:00 Asia/Kuala_Lumpur
    No new features.

    Bugs fixed:

    - RESP3 protocol changes (enabled by default in earlier versions of this module)
    meant that you'd end up with hashrefs or arrayrefs depending on what version
    of Redis you were connecting to. This behaviour is now controlled by the
    `hashrefs` configuration option, and is disabled by default even under RESP3.

3.002     2020-10-12 00:19:11+08:00 Asia/Kuala_Lumpur
    New features:

    - latest commands from 6.0.8, including LMOVE/BLMOVE

3.001     2020-09-22 01:51:28+08:00 Asia/Kuala_Lumpur
    New features:

    - RESP3 no longer establishes a separate client-side caching connection,
    since it can share the main connection
    - `AUTH username password` from latest Redis now supported

    Bugs fixed:

    - proper differentiation between regular arrays and pubsub messages
    when in pubsub mode on a RESP3 connection
    - avoid establishing client-side caching connection until we need one,
    could also have resulted in an attempt to connect to the wrong Redis server
    if using `->configure(host => ...); ->configure(port => ...);`
    - on a failure, continue with next item in pipeline - otherwise subsequent
    requests would get stuck

3.000     2020-09-14 02:00:10+08:00 Asia/Kuala_Lumpur
    New features:

    - RESP3 support, for Redis 6
    - pubsub is now allowed on the same connection as
    other traffic if the connection is in RESP3 mode
    - protocol is autodetected via `HELLO` command,
    pass `protocol => 'resp2'` to disable this

2.007     2020-09-05 02:08:25+08:00 Asia/Kuala_Lumpur
    New features:

    - applies client_name on connection if configured
    - opentracing can be controlled by `->configure(opentracing => 1 || 0)`


lib/Net/Async/Redis/Cluster.pm
package Net::Async::Redis::Cluster;

use strict;
use warnings;

use utf8;

use parent qw(
    Net::Async::Redis::Commands
    IO::Async::Notifier
);

our $VERSION = '3.019'; # VERSION

=encoding utf8

=head1 NAME

Net::Async::Redis::Cluster - support for cluster routing

=head1 SYNOPSIS

 use IO::Async::Loop;
 use Net::Async::Redis::Cluster;
 my $loop = IO::Async::Loop->new;
 $loop->add(
  my $cluster = Net::Async::Redis::Cluster->new(
  )
 );
 await $cluster->bootstrap(
  host => 'redis.local',
 );
 print "Key: " . await $cluster->get('some_key');

=head1 DESCRIPTION

Provides access to a Redis cluster.

Usage is similar to L<Net::Async::Redis> with the addition of a L</bootstrap>
step to find the initial client nodes:

 $loop->add(
  my $cluster = Net::Async::Redis::Cluster->new(
  )
 );
 await $cluster->bootstrap(
  host => 'redis.local',
 );
 print "Key: " . await $cluster->get('some_key');

Note that this adds some overhead to lookups, so you may be better served
by options such as the L<https://github.com/twitter/twemproxy|twemproxy>
proxy routing dæmon, or a service mesh such as L<https://istio.io/|istio>.

=cut

no indirect;
use Syntax::Keyword::Try;
use Future::AsyncAwait;
use Future::Utils qw(fmap_void fmap_concat);
use List::BinarySearch::XS qw(binsearch);
use List::UtilsBy qw(nsort_by);
use List::Util qw(first);
use Digest::CRC qw(crc);
use Cache::LRU;

use Log::Any qw($log);

use Net::Async::Redis;
use Net::Async::Redis::Cluster::Node;

use overload
    '""' => sub { 'NaRedis::Cluster[]' },
    bool => sub { 1 },
    fallback => 1;

use constant MAX_SLOTS => 16384;

BEGIN {
    # It's unlikely that MAX_SLOTS would ever need
    # to change, but it's also important that if we do change it,
    # our bitwise-AND still makes sense in context, i.e. this
    # must always be a power of 2.
    die 'invalid MAX_SLOTS' unless 1 == unpack(
        "b*" => pack(
            N1 => MAX_SLOTS
        )
    ) =~ tr/1/1/;
}

our @CONFIG_KEYS = qw(
    auth
    pipeline_depth
    stream_read_len
    stream_write_len
    on_disconnect
    client_name
    opentracing
    protocol
    hashrefs
    client_side_cache_size
);

=head1 METHODS

=head2 bootstrap

Connects to a Redis server and attempts to discover the cluster node configuration.

Usage:

 try {
  await $cluster->bootstrap(
   host => 'redis-primary.local',
   port => 6379,
  );
 } catch {
  $log->errorf('Unable to bootstrap the Redis cluster: %s', $@);
 }

=cut

async sub bootstrap {
    my ($self, %args) = @_;
    my $redis;
    try {
        $self->add_child(
            $redis = Net::Async::Redis->new(
                host => $args{host},
                port => $args{port},
            )
        );
        await $redis->connect;
        await $self->apply_slots_from_instance($redis);
    } finally {
        $redis->remove_from_parent if $redis;
    }
}

=head2 clientside_cache_events

Provides combined stream of clientside-cache events from all known Redis primary nodes.

=cut

sub clientside_cache_events {
    my ($self) = @_;
    $self->{clientside_cache_events} ||= do {
        $self->ryu->source;
    };
}

=head2 watch_keyspace

L<Net::Async::Redis/watch_keyspace> support for gathering notifications
from all known nodes.

=cut

async sub watch_keyspace {
    my ($self, $pattern) = @_;
    my @sub = await fmap_concat {
        $_->primary_connection->then(sub {
            shift->watch_keyspace($pattern);
        });
    } foreach => [$self->{nodes}->@*], concurrent => 4;

    my $combined = Net::Async::Redis::Subscription->new(
        redis   => $self,
        channel => $pattern
    );

    $combined->events->emit_from(@sub);
    return $combined;
}

=head1 METHODS - Internal

=cut

async sub node_connection_established {
    my ($self, $node, $redis) = @_;
    $self->clientside_cache_events->emit_from($redis->clientside_cache_events) if $redis->is_client_side_cache_enabled;
    return;
}

sub node_config {
    my ($self) = @_;
    return %{$self}{grep exists $self->{$_}, @CONFIG_KEYS};
}

sub configure {
    my ($self, %args) = @_;
    for (@CONFIG_KEYS) {
        $self->{$_} = delete $args{$_} if exists $args{$_};
    }
    die 'invalid protocol requested: ' . $self->{protocol} if defined $self->{protocol} and not $self->{protocol} =~ /^resp[23]$/;

    die 'hashref support requires RESP3 (Redis version 6+)' if defined $self->{protocol} and $self->{protocol} eq 'resp2' and $self->{hashrefs};
    return $self->next::method(%args);
}

sub _init {
    my ($self) = @_;
    $self->{cache} = Cache::LRU->new(
        size => 10_000
    );
}

=head2 hash_slot_for_key

Calculates the CRC16 hash slot for the given key.

Note that keys are expected as bytestrings, if you have a Unicode string
you'd likely want to convert to UTF-8 first.

=cut

sub hash_slot_for_key {
    my ($self, $key) = @_;
    return $self->{cache}->get($key) // do {
        # Extract out the first non-zero-length substring between the first {} character pair,
        # if any...
        my ($hash_string) = $key =~ /^[^\{]*\{([^\}]+)\}/;
        # ... and if we don't have any matching substrings, we just use the full key
        $hash_string //= $key;
        # see Digest::CRC docs and the XMODEM protocol for more details, but essentially:
        # input, width=16, init=0, xor constant for output = 0, reflect output = 0, polynomial = 0x1021,
        # reflect input = 0, continuation = 0
        my $slot = crc($hash_string, 16, 0, 0, 0, 0x1021, 0, 0) & (MAX_SLOTS - 1);
        $self->{cache}->set($key => $slot);
        $slot;
    };
}

=head2 replace_nodes

Swap the existing node configuration out for a new arrayref of nodes.

=cut

sub replace_nodes {
    my ($self, $nodes) = @_;
    delete $self->{slot_cache};
    $_->remove_from_parent for splice $self->{nodes}->@*;
    $self->add_child($_) for $nodes->@*;
    $self->{nodes} = $nodes;
    $self
}

=head2 node_list

Returns a list of the currently-configured nodes.

=cut

sub node_list { shift->{nodes}->@* }

sub slot_cache {
    my ($self) = @_;
    $self->{slot_cache} //= [ (undef) x MAX_SLOTS ];
}

=head2 node_for_slot

Returns the appropriate L<Net::Async::Redis::Cluster::Node|node> for the given hash key (slot).

=cut

sub node_for_slot {
    my ($self, $slot) = @_;
    return $self->slot_cache->[$slot & (MAX_SLOTS - 1)] //= do {
        my @nodes = $self->{nodes}->@*;
        my $idx = binsearch {
            $a < $b->start
            ? -1
            : $a > $b->end
            ? 1
            :0
        } $slot, @nodes;
        $nodes[$idx]
    };
}

async sub connection_for_slot {
    my ($self, $slot) = @_;
    my $node = $self->node_for_slot($slot) or die 'no node found for slot';
    return await $node->primary_connection;
}

sub node_by_host_port {
    my ($self, $host, $port) = @_;
    my $host_port = join ':', $host, $port;
    my ($node) = first { $host_port eq $_->host_port } $self->{nodes}->@*;
    return $node;
}

=head2 register_moved_slot

When we get MOVED error we will use this
sub to rebuild the slot cache

=cut

async sub register_moved_slot {
    my ($self, $slot, $host_port) = @_;
    my ($host, $port) = split /:/, $host_port;
    my $node = $self->node_by_host_port($host, $port);
    unless($node) {
        $log->tracef("Failed to find node %s:%s in the original node list", $host, $port);
        # Has a replica become a primary? Let's look for a valid node
        await fmap_void(async sub {
            my ($node) = @_;
            try {
                my $valid_connection = await $node->primary_connection;
                die 'Cluster status changed and cannot find a valid information source' unless $valid_connection;
                # We'll let *all* our nodes try to tell us about slots, the operation
                # should be atomic so whichever one(s) succeed are hopefully consistent
                await $self->apply_slots_from_instance($valid_connection);
            } catch {
                $log->tracef("Node at %s was invalid", $node->primary);
            }
        }, concurrent => 4, foreach => [ $self->{nodes}->@* ]);
        # Try again else propgate failure
        $node = $self->node_by_host_port($host, $port)
            or die "Slot $slot has been moved to unknown node";
    }
    $self->slot_cache->[$slot & (MAX_SLOTS - 1)] = $node;
    return $node;
}

=head2 apply_slots_from_instance

Connect to a random instance in the cluster
and execute CLUSTER SLOTS to get information
about the slots and their distribution.

=cut

async sub apply_slots_from_instance {
    my ($self, $redis) = @_;
    my ($slots) = await $redis->cluster_slots;
    $log->tracef('Have %d slots', 0 + @$slots);

    my @nodes;
    for my $slot_data (nsort_by { $_->[0] } $slots->@*) {
        my $node = Net::Async::Redis::Cluster::Node->from_arrayref(
            $slot_data,
            cluster => $self,
            $self->node_config
        );
        $log->tracef(
            'Node %s (%s) handles slots %d-%d and has %d replica(s) - %s',
            $node->id,
            $node->primary,
            $node->start,
            $node->end,
            $node->replica_count,
            $node
        );

        push @nodes, $node;
    }

    $self->replace_nodes(\@nodes);
}

=head2 execute_command

Lookup the correct node for the key then execute the command on that node,
if there is a mismatch between our slot hashes and Redis's hashes
we will attempt to rebuild the slot hashes and try again

=cut

async sub execute_command {
    my ($self, @cmd) = @_;
    $log->tracef('Will execute %s on cluster', join(' ', @cmd));
    my $k;
    if($cmd[0] eq 'XREADGROUP' or $cmd[0] eq 'XREAD') {
        my ($idx) = grep { $cmd[$_] eq 'STREAMS' } 0..$#cmd;
        $k = $cmd[$idx + 1];
    } else {
        # So far our longest command name is 2 words
        my $key_idx = $Net::Async::Redis::Commands::KEY_FINDER{$cmd[0]};
        $key_idx //= $Net::Async::Redis::Commands::KEY_FINDER{$cmd[0] . ' ' . $cmd[1]} if @cmd > 1;
        die 'no index found for ' . join(' ', @cmd) unless defined $key_idx;
        $k = $cmd[$key_idx];
    }
    my $slot = $self->hash_slot_for_key($k);
    $log->tracef('Look up hash slot for %s - %d', $k, $slot);
    my $redis = await $self->connection_for_slot($slot);
    # Some commands have modifiers around them for RESP2/3 transparent support
    my ($command, @args) = @cmd;
    try {
        $command = lc $command;
        return await $redis->$command(@args);
    } catch ($e) {
        die $e unless $e =~ /MOVED/;
        my ($moved, $slot, $host_port) = split ' ', $e;
        await $self->register_moved_slot($slot => $host_port);
        return await $self->execute_command(@cmd);
    }
}

=head2 ryu

A L<Ryu::Async> instance for source/sink creation.

=cut

sub ryu {
    my ($self) = @_;
    $self->{ryu} ||= do {
        $self->add_child(
            my $ryu = Ryu::Async->new
        );
        $ryu
    }
}

1;

=head1 AUTHOR

Tom Molesworth C<< <TEAM@cpan.org> >> plus contributors as mentioned in
L<Net::Async::Redis/CONTRIBUTORS>.

=head1 LICENSE

Copyright Tom Molesworth and others 2015-2022.
Licensed under the same terms as Perl itself.

